Bottom: d77df4a919f8423532cd3c1855497cc178299721
Top:    3473a40e0056c1f3264c3c5be0213424ec97924c
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2014-07-04 01:45:02 +0100

hid/*: layer_sort()  Only copper layers participate in the layer stack

Don't consider the layers greater than or equal to max_copper_layer,
ie.. Silk layers are not in the layer stack. Actually, don't even
worry about checking the layer is a copper layer - it will be.

Use GetLayerGroupNumberByNumber() directly, there is no need to worry
about being passed strange layer numbers.

Rename layer_sort() to layer_stack_sort(), reflecting our (new) preconditions.


---

diff --git a/src/hid/gerber/gerber.c b/src/hid/gerber/gerber.c
index 29691bd..31b8c90 100644
--- a/src/hid/gerber/gerber.c
+++ b/src/hid/gerber/gerber.c
@@ -411,23 +411,17 @@ gerber_get_export_options (int *n)
 }
 
 static int
-group_for_layer (int l)
+layer_stack_sort (const void *va, const void *vb)
 {
-  if (l < max_copper_layer + 2 && l >= 0)
-    return GetLayerGroupNumberByNumber (l);
-  /* else something unique */
-  return max_group + 3 + l;
-}
+  int a_layer = *(int *) va;
+  int b_layer = *(int *) vb;
+  int a_group = GetLayerGroupNumberByNumber (a_layer);
+  int b_group = GetLayerGroupNumberByNumber (b_layer);
 
-static int
-layer_sort (const void *va, const void *vb)
-{
-  int a = *(int *) va;
-  int b = *(int *) vb;
-  int d = group_for_layer (b) - group_for_layer (a);
-  if (d)
-    return d;
-  return b - a;
+  if (b_group != a_group)
+    return b_group - a_group;
+
+  return b_layer - a_layer;
 }
 
 static void
@@ -670,7 +664,7 @@ gerber_do_export (HID_Attr_Val * options)
       print_layer[i] = 1;
 
   memcpy (saved_layer_stack, LayerStack, sizeof (LayerStack));
-  qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_sort);
+  qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_stack_sort);
   linewidth = -1;
   lastcap = -1;
   lastgroup = -1;
diff --git a/src/hid/png/png.c b/src/hid/png/png.c
index 0c68dad..2f1590b 100644
--- a/src/hid/png/png.c
+++ b/src/hid/png/png.c
@@ -505,33 +505,22 @@ png_get_export_options (int *n)
 static int top_group, bottom_group;
 
 static int
-group_for_layer (int l)
+layer_stack_sort (const void *va, const void *vb)
 {
-  if (l < max_copper_layer + 2 && l >= 0)
-    return GetLayerGroupNumberByNumber (l);
-  /* else something unique */
-  return max_group + 3 + l;
-}
+  int a_layer = *(int *) va;
+  int b_layer = *(int *) vb;
+  int a_group = GetLayerGroupNumberByNumber (a_layer);
+  int b_group = GetLayerGroupNumberByNumber (b_layer);
+  int aside = (a_group == bottom_group ? 0 : a_group == top_group ? 2 : 1);
+  int bside = (b_group == bottom_group ? 0 : b_group == top_group ? 2 : 1);
 
-static int
-layer_sort (const void *va, const void *vb)
-{
-  int a = *(int *) va;
-  int b = *(int *) vb;
-  int al = group_for_layer (a);
-  int bl = group_for_layer (b);
-  int d = bl - al;
+  if (bside != aside)
+    return bside - aside;
 
-  if (a >= 0 && a <= max_copper_layer + 1)
-    {
-      int aside = (al == bottom_group ? 0 : al == top_group ? 2 : 1);
-      int bside = (bl == bottom_group ? 0 : bl == top_group ? 2 : 1);
-      if (bside != aside)
-	return bside - aside;
-    }
-  if (d)
-    return d;
-  return b - a;
+  if (b_group != a_group)
+    return b_group - a_group;
+
+  return b_layer - a_layer;
 }
 
 static const char *filename;
@@ -601,7 +590,7 @@ png_hid_export_to_file (FILE * the_file, HID_Attr_Val * options)
 
       top_group = GetLayerGroupNumberBySide (TOP_SIDE);
       bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
-      qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_sort);
+      qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_stack_sort);
 
       CLEAR_FLAG(THINDRAWFLAG, PCB);
       CLEAR_FLAG(THINDRAWPOLYFLAG, PCB);
diff --git a/src/hid/ps/eps.c b/src/hid/ps/eps.c
index cf63487..cff21e3 100644
--- a/src/hid/ps/eps.c
+++ b/src/hid/ps/eps.c
@@ -152,33 +152,22 @@ eps_get_export_options (int *n)
 static int top_group, bottom_group;
 
 static int
-group_for_layer (int l)
+layer_stack_sort (const void *va, const void *vb)
 {
-  if (l < max_copper_layer + 2 && l >= 0)
-    return GetLayerGroupNumberByNumber (l);
-  /* else something unique */
-  return max_group + 3 + l;
-}
+  int a_layer = *(int *) va;
+  int b_layer = *(int *) vb;
+  int a_group = GetLayerGroupNumberByNumber (a_layer);
+  int b_group = GetLayerGroupNumberByNumber (b_layer);
+  int aside = (a_group == bottom_group ? 0 : a_group == top_group ? 2 : 1);
+  int bside = (b_group == bottom_group ? 0 : b_group == top_group ? 2 : 1);
 
-static int
-layer_sort (const void *va, const void *vb)
-{
-  int a = *(int *) va;
-  int b = *(int *) vb;
-  int al = group_for_layer (a);
-  int bl = group_for_layer (b);
-  int d = bl - al;
+  if (bside != aside)
+    return bside - aside;
 
-  if (a >= 0 && a <= max_copper_layer + 1)
-    {
-      int aside = (al == bottom_group ? 0 : al == top_group ? 2 : 1);
-      int bside = (bl == bottom_group ? 0 : bl == top_group ? 2 : 1);
-      if (bside != aside)
-	return bside - aside;
-    }
-  if (d)
-    return d;
-  return b - a;
+  if (b_group != a_group)
+    return b_group - a_group;
+
+  return b_layer - a_layer;
 }
 
 static const char *filename;
@@ -261,7 +250,7 @@ eps_hid_export_to_file (FILE * the_file, HID_Attr_Val * options)
     {
       top_group = GetLayerGroupNumberBySide (TOP_SIDE);
       bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
-      qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_sort);
+      qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_stack_sort);
     }
   fprintf (f, "%%!PS-Adobe-3.0 EPSF-3.0\n");
   linewidth = -1;
diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index 8e3a2ce..690697a 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -417,23 +417,17 @@ ps_get_export_options (int *n)
 }
 
 static int
-group_for_layer (int l)
+layer_stack_sort (const void *va, const void *vb)
 {
-  if (l < max_copper_layer + 2 && l >= 0)
-    return GetLayerGroupNumberByNumber (l);
-  /* else something unique */
-  return max_group + 3 + l;
-}
+  int a_layer = *(int *) va;
+  int b_layer = *(int *) vb;
+  int a_group = GetLayerGroupNumberByNumber (a_layer);
+  int b_group = GetLayerGroupNumberByNumber (b_layer);
 
-static int
-layer_sort (const void *va, const void *vb)
-{
-  int a = *(int *) va;
-  int b = *(int *) vb;
-  int d = group_for_layer (b) - group_for_layer (a);
-  if (d)
-    return d;
-  return b - a;
+  if (b_group != a_group)
+    return b_group - a_group;
+
+  return b_layer - a_layer;
 }
 
 void
@@ -671,7 +665,7 @@ ps_hid_export_to_file (FILE * the_file, HID_Attr_Val * options)
       global.print_layer[i] = 1;
 
   memcpy (saved_layer_stack, LayerStack, sizeof (LayerStack));
-  qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_sort);
+  qsort (LayerStack, max_copper_layer, sizeof (LayerStack[0]), layer_stack_sort);
 
   global.linewidth = -1;
   /* reset static vars */
